### PR TITLE
serializer: Recognizer interface should take a buffer

### PR DESCRIPTION
**Pull request by smarterclayton**
The original intent of the Recognizer interface was to take a stream and peek into the contents. Over time all of the callers migrated to instead performing their own peek and then passing the buffer as a bytes.NewBuffer to the interface as a stream, which was then implemented as a read on the other side. This resulted in an excess buffer allocation in general decoder usage.  Recognizer is used in the apiserver deserialization path from etcd.

Update the interface to take a []bytes slice and change all callers to stop creating buffered readers, then update the JSON serializer to check the buffer directly for "json-ness". Avoids a 2kb allocation once per recognizer Decode()

/kind cleanup

```release-note
NONE
```

```docs
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go [hunk 1]
```diff
@@ -120,7 +120,7 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 
 	if intoUnknown, ok := into.(*runtime.Unknown); ok && intoUnknown != nil {
 		*intoUnknown = unk
-		if ok, _, _ := s.RecognizesData(bytes.NewBuffer(unk.Raw)); ok {
+		if ok, _, _ := s.RecognizesData(unk.Raw); ok {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:123]: cc @enj

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 2]
```diff
@@ -343,6 +343,12 @@ func GuessJSONStream(r io.Reader, size int) (io.Reader, []byte, bool) {
 	return buffer, b, hasJSONPrefix(b)
 }
 
+// IsJSONBuffer scans the provided buffer, looking
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:346]: this is no more correct than `GuessJSONStream` and still fails https://github.com/kubernetes/kubernetes/issues/78946

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:123]: @smarterclayton pretty sure the existing code is just wrong, see #98439.

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:346]: I'm not in any way addressing those other bugs.  This change isn't fixing known gaps with guess.

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go (hunk 1 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:123]: Either way doesn't have anything to do with this change and is orthogonal (I will review that separately).

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:346]: (i'm all for improving guess as efficiently as possible, which means finding all callers and having them all consistently use a bigger buffer, but there is no solution other than reading it all into memory to get perfection)

**Comment by liggitt** [comment]
making RecognizingDecoder coherent with `runtime.Decoder` (which also takes []byte) makes sense

/lgtm
/approve

**Pull request merged**